src/python/py/models/builder.py

```diff
@@ -1074,25 +1143,127 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
+            w2_list.append(torch.reshape(bsm.experts[i].w2.weight, (inter_size, hidden_size)))
+            w3_list.append(torch.reshape(bsm.experts[i].w3.weight, (hidden_size, inter_size)))
+
+        moe_expert_1_name = f"model.layers.{layer_id}.moe.weight_1"
+        moe_expert_2_name = f"model.layers.{layer_id}.moe.weight_2"
+        moe_expert_3_name = f"model.layers.{layer_id}.moe.weight_3"
+
+        moe_experts_weight1_full = torch.stack(w1_list, dim=0).detach().numpy()
+        moe_experts_weight2_full = torch.stack(w2_list, dim=0).detach().numpy()
+        moe_experts_weight3_full = torch.stack(w3_list, dim=0).detach().numpy()
+
+        moe_experts_weight1 = moe_experts_weight1_full
+        moe_experts_weight2 = moe_experts_weight2_full
+        moe_experts_weight3 = moe_experts_weight3_full
+
+        if self.world_size > 1:
+            def get_fc1_tensor_shards(expert_weights):
+                return (
+                    expert_weights.reshape(-1, inter_size, hidden_size)
+                    .transpose(0, 2, 1)[
+                        :, :, self.rank * inter_size // self.world_size : (self.rank + 1) * inter_size // self.world_size
+                    ]
+                    .transpose(0, 2, 1).reshape(-1, hidden_size, inter_size // self.world_size)
+                )
+
+            def get_fc2_tensor_shards(expert_weights):
+                return (
+                    expert_weights.reshape(-1, hidden_size, inter_size)
+                    .transpose(0, 2, 1)[
+                        :, self.rank * inter_size // self.world_size : (self.rank + 1) * inter_size // self.world_size, :
+                    ]
+                    .transpose(0, 2, 1).reshape(-1, inter_size // self.world_size, hidden_size)
+                )
+
+            moe_experts_weight1 = get_fc1_tensor_shards(moe_experts_weight1_full)
+            moe_experts_weight2 = get_fc2_tensor_shards(moe_experts_weight2_full)
+            moe_experts_weight3 = get_fc1_tensor_shards(moe_experts_weight3_full)
+
+        self.make_external_tensor(moe_experts_weight1.astype(self.to_numpy_dtype[self.io_dtype]), moe_expert_1_name)
+        self.make_external_tensor(moe_experts_weight2.astype(self.to_numpy_dtype[self.io_dtype]), moe_expert_2_name)
+        self.make_external_tensor(moe_experts_weight3.astype(self.to_numpy_dtype[self.io_dtype]), moe_expert_3_name)
+
+        bias_ph = "" # Placeholder for bias
+        inputs = [root_input, f"{gate_reshape_name}/output_0", moe_expert_1_name, bias_ph, moe_expert_2_name, bias_ph, moe_expert_3_name]
+        output = f"{moe_name}/output_0"
+        if self.world_size > 1:        
+            self.make_node("ShardedMoE", inputs=inputs, outputs=[output], name=moe_name, domain="com.microsoft",
+                            k=top_k, activation_type=activation_type, normalize_routing_weights=normalize_routing_weights, tensor_shards=self.world_size)
```

Comment:
Can the logic to create an MoE node be factored out into one node creation function? Something like this:

```
op_type = f"{'Sharded' if self.world_size > 1 else ''}MoE"
kwargs = {"tensor_shards": self.world_size} if self.world_size > 1 else {}
self.make_node(
    op_type, inputs=inputs, outputs=[output], name=moe_name, domain="com.microsoft",
    k=top_k, activation_type=activation_type, normalize_routing_weights=normalize_routing_weights,
    **kwargs,
)
```